### PR TITLE
Enhances the set of cases 'use pattern matching' works for.  

### PR DESCRIPTION
**Pull request by CyrusNajmabadi**
Fixes https://github.com/dotnet/roslyn/issues/65308

Example of what this now improves:

image

Going to test this on Roslyn.sln to see if it encounters any issues.

### CODE REVIEW

File: src/Analyzers/CSharp/Tests/UsePatternMatching/CSharpAsAndMemberAccessTests.cs [hunk 1]
```diff
@@ -0,0 +1,505 @@
...
+                            }
+                        }
+                    }
+                    """,
+                FixedCode = """
+                    class C
+                    {
+                        void M(object o)
+                        {
+                            if (o is string { Length: not 0 })
```
Comment (CyrusNajmabadi) [inline src/Analyzers/CSharp/Tests/UsePatternMatching/CSharpAsAndMemberAccessTests.cs:183]: this is incorrect.  working on fix now.

**Comment by CyrusNajmabadi** [comment]
@genlu @akhera99 @ryzngard ptal. Thanks!

File: src/Analyzers/CSharp/Analyzers/UsePatternMatching/UsePatternMatchingHelpers.cs [hunk 2]
```diff
@@ -0,0 +1,94 @@
...
+            out BinaryExpressionSyntax? binaryExpression,
+            out IsPatternExpressionSyntax? isPatternExpression,
+            out LanguageVersion requiredLanguageVersion)
+        {
+            conditionalAccessExpression = null;
+            binaryExpression = null;
+            isPatternExpression = null;
+            requiredLanguageVersion = LanguageVersion.CSharp8;
+
+            if (asExpression.Kind() == SyntaxKind.AsExpression)
```
Comment (dibarbet) [inline src/Analyzers/CSharp/Analyzers/UsePatternMatching/UsePatternMatchingHelpers.cs:26]: nit - maybe
```
if (asExpression.Kind() != SyntaxKind.AsExpression)
{
   return false;
}
```

File: src/Analyzers/Core/Analyzers/EnforceOnBuildValues.cs [hunk 3]
```diff
@@ -85,6 +85,7 @@ internal static class EnforceOnBuildValues
         public const EnforceOnBuild RemoveRedundantNullableDirective = /*IDE0240*/ EnforceOnBuild.Recommended;
         public const EnforceOnBuild RemoveUnnecessaryNullableDirective = /*IDE0241*/ EnforceOnBuild.Recommended;
         public const EnforceOnBuild MakeStructReadOnly = /*IDE0250*/ EnforceOnBuild.Recommended;
+        public const EnforceOnBuild UsePatternMatchingAsAndMemberAccess = /*IDE0260*/ EnforceOnBuild.Recommended;
```
Comment (dibarbet) [inline src/Analyzers/Core/Analyzers/EnforceOnBuildValues.cs:88]: I am not familiar with this list - why do we need an explicit set of analyzers that are enforced on build?

**Review by dibarbet — APPROVED** [review]

File: src/Analyzers/CSharp/Analyzers/UsePatternMatching/CSharpAsAndMemberAccessDiagnosticAnalyzer.cs [hunk 4]
```diff
@@ -0,0 +1,182 @@
...
+            var semanticModel = context.SemanticModel;
+            var asExpression = (BinaryExpressionSyntax)context.Node;
+
+            if (!UsePatternMatchingHelpers.TryGetPartsOfAsAndMemberAccessCheck(
+                    asExpression, out var conditionalAccessExpression, out var binaryExpression, out var isPatternExpression, out var requiredLanguageVersion))
+            {
+                return;
+            }
+
+            if (context.Compilation.LanguageVersion() < requiredLanguageVersion)
```
[the remark this answers is not in the document]
Comment (CyrusNajmabadi) [reply src/Analyzers/CSharp/Analyzers/UsePatternMatching/CSharpAsAndMemberAccessDiagnosticAnalyzer.cs:78]: the 'if' is waht returns the 'requiredLanguageVersion'.  Basically there are different types of fixes we can provide depending on the user code, and the different fixes have different min-lang-versions supported.  So we do the syntactic analysis, see what the fix would need and bail out if we can't provide it.

File: src/Analyzers/CSharp/Analyzers/UsePatternMatching/UsePatternMatchingHelpers.cs (hunk 2 above)
Comment (CyrusNajmabadi) [reply src/Analyzers/CSharp/Analyzers/UsePatternMatching/UsePatternMatchingHelpers.cs:26]: yeah, that's better.  will do in followup :)

File: src/Analyzers/CSharp/Analyzers/UsePatternMatching/CSharpAsAndMemberAccessDiagnosticAnalyzer.cs (hunk 4 above)
[the remark this answers is not in the document]
Comment (dibarbet) [reply src/Analyzers/CSharp/Analyzers/UsePatternMatching/CSharpAsAndMemberAccessDiagnosticAnalyzer.cs:78]: ah oops missed that

**Pull request merged**